core/embed/rust/src/ui/model_tr/component/button_controller.rs

```diff
@@ -0,0 +1,498 @@
+use super::{
+    theme, Button, ButtonDetails, ButtonLayout, ButtonPos, HoldToConfirm, HoldToConfirmMsg,
+    LoaderStyleSheet,
+};
+use crate::{
+    time::Duration,
+    ui::{
+        component::{base::Event, Component, EventCtx, Pad},
+        event::{ButtonEvent, PhysicalButton},
+        geometry::Rect,
+    },
+};
+
+/// All possible states buttons (left and right) can be at.
+#[derive(Copy, Clone, PartialEq, Eq)]
+enum ButtonState {
+    /// Both buttons are in untouched state.
+    /// _ _
+    /// NEXT: OneDown
+    Nothing,
+    /// One Button is down when previously nothing was.
+    /// _ _  ... ↓ _ | _ ↓
+    /// NEXT: Nothing, BothDown
+    OneDown(PhysicalButton),
+    /// Both buttons are down ("middle-click").
+    /// ↓ _ | _ ↓ ... ↓ ↓
+    /// NEXT: OneReleased
+    BothDown,
+    /// One button is down when previously both were.
+    /// Happens when "middle-click" is performed.
+    /// ↓ ↓ ... ↓ _ | _ ↓
+    /// NEXT: Nothing, BothDown
+    OneReleased(PhysicalButton),
+    /// One button is down after it triggered a HoldToConfirm event.
+    /// Needed so that we can cleanly reset the state.
+    /// ↓ _ | _ ↓ ... ↓ _ | _ ↓
+    /// NEXT: Nothing
+    HTCNeedsRelease(PhysicalButton),
+}
+
+pub enum ButtonControllerMsg {
+    Pressed(ButtonPos),
+    Triggered(ButtonPos),
+}
+
+/// Defines what kind of button should be currently used.
+pub enum ButtonType<T>
+where
+    T: AsRef<str> + Clone + From<&'static str>,
+{
+    Button(Button<T>),
+    HoldToConfirm(HoldToConfirm<T>),
+    Nothing,
+}
+
+impl<T> ButtonType<T>
+where
+    T: AsRef<str> + Clone + From<&'static str>,
+{
+    pub fn from_button_details(pos: ButtonPos, btn_details: Option<ButtonDetails<T>>) -> Self {
+        if let Some(btn_details) = btn_details {
+            if btn_details.duration.is_some() {
+                Self::HoldToConfirm(Self::get_hold_to_confirm(pos, btn_details))
+            } else {
+                Self::Button(Self::get_button(pos, btn_details))
+            }
+        } else {
+            Self::Nothing
+        }
+    }
+
+    /// Create `Button` component from `btn_details`.
+    fn get_button(pos: ButtonPos, btn_details: ButtonDetails<T>) -> Button<T> {
+        // Deciding between text and icon
+
+        if let Some(text) = btn_details.text.clone() {
+            Button::with_text(pos, text, btn_details.style())
+        } else if let Some(icon) = btn_details.icon {
+            Button::with_icon(pos, icon, btn_details.style())
+        } else {
+            panic!("ButtonContainer: no text or icon provided");
+        }
+    }
```

Comment:
This looks more like `Button::from_details(pos: ButtonPos, btn_details: ButtonDetails<T>) -> Self`, no?

Reply:
Right, corrected in [bca3e8b](https://github.com/trezor/trezor-firmware/pull/2610/commits/bca3e8b694855483a317d468230324561acb3ba9) together with some smaller Button improvements